oauthproxy.go

```diff
@@ -961,15 +961,57 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
 //
 //nolint:gosimple
 func authOnlyAuthorize(req *http.Request, s *sessionsapi.SessionState) bool {
-	// Allow secondary group restrictions based on the `allowed_groups`
-	// querystring parameter
-	if !checkAllowedGroups(req, s) {
-		return false
+	constraints := []func(*http.Request, *sessionsapi.SessionState) bool{
+		checkAllowedGroups,
+		checkAllowedEmailDomains,
+	}
+
+	for _, constraint := range constraints {
+		if !constraint(req, s) {
+			return false
+		}
 	}
 
 	return true
 }
 
+// checkAllowedEmailDomains allow email domain restrictions based on the `allowed_email_domains`
+// querystring parameter
+func checkAllowedEmailDomains(req *http.Request, s *sessionsapi.SessionState) bool {
+	allowedEmailDomains := extractAllowedEmailDomains(req)
+	if len(allowedEmailDomains) == 0 {
+		return true
+	}
+
+	emailSplited := strings.Split(s.Email, "@")
+	if len(emailSplited) != 2 {
+		return false
+	}
+
+	if _, ok := allowedEmailDomains[emailSplited[1]]; ok {
+		return true
+	}
+
+	return false
+}
+
+func extractAllowedEmailDomains(req *http.Request) map[string]struct{} {
```

Comment:
This and `extractAllowedGroups` are nearly identical -- maybe we should generalize it and use the querystring param as an extra argument (appears to be the only difference)?

Reply:
Refactored